### PR TITLE
adding non-shared BLSTM layers in CTC

### PR DESCRIPTION
**Pull request by sw005320**

### CODE REVIEW

**Comment by sw005320** [comment]
I'm observing some small improvement when I added a non-shared BLSTM layer before the CTC objective.

File: src/bin/asr_train.py [hunk 1]
```diff
@@ -63,10 +63,16 @@ def main():
...
                         help='Subsample input frames x_y_z means subsample every x frame at 1st layer, '
                              'every y frame at 2nd layer etc.')
-    # loss
-    parser.add_argument('--ctc_type', default='warpctc', type=str,
-                        choices=['chainer', 'warpctc'],
-                        help='Type of CTC implementation to calculate loss.')
+    # ctc
+    parser.add_argument('--ctype', default='warpctc', type=str,
+                        choices=['warpctc', 'chainer', 'blstm'],
+                        help='Type of CTC architecture.')
```
Comment (kan-bayashi) [inline src/bin/asr_train.py]: The name `blstm` is a little bit confusing.
How about adding non-shared layer when `clayers > 0` while keeping `ctype=warpctc`?

File: src/nets/e2e_asr_attctc.py [hunk 2]
```diff
@@ -331,6 +337,78 @@ def log_softmax(self, hs):
...
+        :return:
+        '''
+        self.loss = None
+        ilens = [x.shape[0] for x in hs]
+        olens = [x.shape[0] for x in ys]
+
+        hy, cy, hs = self.nblstm(None, None, hs)
+        hs = self.l_last(F.vstack(hs))  # (sum _utt frame_utt) x dim
+        hs = F.split_axis(hs, np.cumsum(ilens[:-1]), axis=0)
+        del hy, cy
```
Comment (kan-bayashi) [inline src/nets/e2e_asr_attctc.py]: `_, _, hs = self.nblstm(None, None, hs)`

File: src/nets/e2e_asr_attctc.py [hunk 3]
```diff
@@ -331,6 +337,78 @@ def log_softmax(self, hs):
...
+    def log_softmax(self, hs, hlen):
+        '''log_softmax of frame activations
+
+        :param hs:
+        :return:
+        '''
+        hy, cy, hs = self.nblstm(None, None, hs)
+        hs = self.l_last(F.vstack(hs))  # (sum _utt frame_utt) x dim
+        hs = F.split_axis(hs, np.cumsum(hlen[:-1]), axis=0)
+        del hy, cy
```
Comment (kan-bayashi) [inline src/nets/e2e_asr_attctc.py]: _, _, hs = self.nblstm(None, None, hs)

File: src/nets/e2e_asr_attctc_th.py [hunk 4]
```diff
@@ -435,6 +444,78 @@ def log_softmax(self, hpad):
...
+
+        :param hs:
+        :param ys:
+        :return:
+        '''
+        self.loss = None
+
+        hpack = pack_padded_sequence(hpad, ilens, batch_first=True)
+        hs, (hy, cy) = self.nblstm(hpack)
+        del hy, cy
```
Comment (kan-bayashi) [inline src/nets/e2e_asr_attctc_th.py]: hs, _ = self.nblstm(hpack)

File: src/nets/e2e_asr_attctc_th.py [hunk 5]
```diff
@@ -435,6 +444,78 @@ def log_softmax(self, hpad):
...
+    def log_softmax(self, hpad, ilens):
+        '''log_softmax of frame activations
+
+        :param hs:
+        :return:
+        '''
+
+        hpack = pack_padded_sequence(hpad, ilens, batch_first=True)
+        hs, (hy, cy) = self.nblstm(hpack)
+        del hy, cy
```
Comment (kan-bayashi) [inline src/nets/e2e_asr_attctc_th.py]: hs, _ = self.nblstm(hpack)

**Comment by kan-bayashi** [comment]
LGTM
But if you want to make more coherent, please make it to be compatible with `clayers` options in using CTC of chainer.

**Pull request closed without merging**